### PR TITLE
Remove asserts from non tests

### PR DESCRIPTION
**Pull request by swamper123**
Fix for #215

### CODE REVIEW

File: snap7/client.py [hunk 1]
```diff
@@ -244,8 +244,9 @@ def read_area(self, area: str, dbnumber: int, start: int, size: int) -> bytearra
         :param start: offset to start writing
         :param size: number of units to read
         """
-        assert area in snap7.types.areas.values()
-        if area == snap7.types.S7AreaTM:
+        if area not in snap7.types.areas.values():
+            raise NotImplementedError(f"{area} is not implemented in snap7.types")
```
Comment (gijzelaerr) [inline snap7/client.py:248]: i think this should be a type error

File: snap7/client.py [hunk 2]
```diff
@@ -367,7 +369,8 @@ def set_connection_params(self, address: str, local_tsap: int, remote_tsap: int)
         :param local_tsap: Local TSAP (PC TSAP)
         :param remote_tsap: Remote TSAP (PLC TSAP)
         """
-        assert re.match(ipv4, address), f'{address} is invalid ipv4'
+        if re.match(ipv4, address) is False:
```
Comment (gijzelaerr) [inline snap7/client.py:372]: if not re.match

File: snap7/common.py [hunk 3]
```diff
@@ -73,7 +73,8 @@ def error_text(error, context: str = "client") -> bytes:
     :param context: server, client or partner
     :returns: the error string
     """
-    assert context in ("client", "server", "partner")
+    if context not in ("client", "server", "partner"):
+        raise NotImplementedError(f"Unkown context {context} used, should be either client, server or partner")
```
Comment (gijzelaerr) [inline snap7/common.py:77]: type error

File: snap7/server.py [hunk 4]
```diff
@@ -242,7 +243,8 @@ def set_mask(self, kind: int, mask: int):
     def set_cpu_status(self, status: int):
         """Sets the Virtual CPU status.
         """
-        assert status in snap7.types.cpu_statuses, f'unknown cpu state {status}'
+        if status not in snap7.types.cpu_statuses:
+            raise KeyError(f"unknown cpu state {status}")
```
Comment (gijzelaerr) [inline snap7/server.py:247]: value error

File: snap7/util.py [hunk 5]
```diff
@@ -200,7 +202,8 @@ def set_string(bytearray_: bytearray, byte_index: int, value: str, max_size: int
     :params value: string data
     :params max_size: max possible string size
     """
-    assert isinstance(value, str)
+    if isinstance(value, str) is False:
```
Comment (gijzelaerr) [inline snap7/util.py:205]: if not

File: snap7/util.py [hunk 6]
```diff
@@ -106,7 +106,9 @@ def set_bool(bytearray_: bytearray, byte_index: int, bool_index: int, value: boo
     """
     Set boolean value on location in bytearray
     """
-    assert value in [0, 1, True, False]
+    if isinstance(value, bool) is False:
```
Comment (gijzelaerr) [inline snap7/util.py:109]: if not

File: snap7/util.py [hunk 7]
```diff
@@ -530,12 +535,14 @@ def __getitem__(self, key):
         """
         Get a specific db field
         """
-        assert key in self._specification
+        if key not in self._specification:
+            raise KeyError(f"Value {key} was not found in specifications.")
```
Comment (gijzelaerr) [inline snap7/util.py:539]: this is already implicitly done by just accessing the key

File: snap7/util.py [hunk 8]
```diff
@@ -530,12 +535,14 @@ def __getitem__(self, key):
...
         index, _type = self._specification[key]
         return self.get_value(index, _type)
 
     def __setitem__(self, key, value):
-        assert key in self._specification
+        if  key not in self._specification:
+            raise KeyError(f"Value {key} was not found in specifications.")
```
Comment (gijzelaerr) [inline snap7/util.py:545]: this is already implicitly done by just accessing the key

**Review by gijzelaerr — CHANGES_REQUESTED** [review]

File: snap7/server.py (hunk 4 above)
Comment (swamper123) [reply snap7/server.py:247]: cpu_statuses is a dict, so I guess it is implicit done anyway

File: snap7/server.py [hunk 9]
```diff
@@ -242,7 +244,8 @@ def set_mask(self, kind: int, mask: int):
     def set_cpu_status(self, status: int):
         """Sets the Virtual CPU status.
         """
-        assert status in snap7.types.cpu_statuses, f'unknown cpu state {status}'
+        if status not in snap7.types.cpu_statuses:
+            raise Snap7Exception(f"The cpu state ({status}) is invalid")
```
Comment (gijzelaerr) [inline snap7/server.py:248]: sorry, i think we should probably just get rid of this snap7exection. I think this is a value error.

File: snap7/client.py [hunk 10]
```diff
@@ -620,8 +623,8 @@ def wait_as_completion(self, timeout: int) -> int:
 
     def _prepare_as_read_area(self, area: str, size: int) -> Tuple[int, Array]:
         if area not in snap7.types.areas.values():
-            raise NotImplementedError(f"{area} is not implemented in snap7.types")
-        if area == snap7.types.S7AreaTM:
+            raise Snap7Exception(f"{area} is not implemented in snap7.types")
```
Comment (gijzelaerr) [inline snap7/client.py:626]: sorry, i think we should probably just get rid of this snap7exection. I think this is a value error.

File: snap7/client.py [hunk 11]
```diff
@@ -649,8 +652,8 @@ def as_read_area(self, area: str, dbnumber: int, start: int, size: int, wordlen:
 
     def _prepare_as_write_area(self, area: str, data: bytearray) -> Tuple[int, Array]:
         if area not in snap7.types.areas.values():
-            raise NotImplementedError(f"{area} is not implemented in snap7.types")
-        if area == snap7.types.S7AreaTM:
+            raise Snap7Exception(f"{area} is not implemented in snap7.types")
```
Comment (gijzelaerr) [inline snap7/client.py:655]: sorry, i think we should probably just get rid of this snap7exection. I think this is a value error.

**Review by gijzelaerr — CHANGES_REQUESTED** [review]

File: snap7/client.py [hunk 12]
```diff
@@ -244,8 +244,9 @@ def read_area(self, area: str, dbnumber: int, start: int, size: int) -> bytearra
         :param start: offset to start writing
         :param size: number of units to read
         """
-        assert area in snap7.types.areas.values()
-        if area == snap7.types.S7AreaTM:
+        if area not in snap7.types.areas.values():
+            raise Snap7Exception(f"{area} is not implemented in snap7.types")
```
Comment (gijzelaerr) [inline snap7/client.py:248]: one left jere.

File: snap7/server.py [hunk 13]
```diff
@@ -10,6 +10,7 @@
 import snap7
 import snap7.types
 from snap7.common import check_error, load_library, ipv4
+from snap7.exceptions import Snap7Exception
```
Comment (gijzelaerr) [inline snap7/server.py:13]: this import can be removed

File: test/test_server.py [hunk 14]
```diff
@@ -6,6 +6,7 @@
 import snap7.error
 import snap7.server
 import snap7.types
+from snap7.exceptions import Snap7Exception
```
Comment (gijzelaerr) [inline test/test_server.py:9]: this import can ben removed.

**Review by gijzelaerr — CHANGES_REQUESTED** [review]

**Review by gijzelaerr — APPROVED** [review]

**Pull request merged**